File: hackerrank/problem solving/algorithms/2 - implementation/40 - acm icpc team/acm_icpc_team.py
# https://www.hackerrank.com/challenges/acm-icpc-team/problem?isFullScreen=true

from itertools import combinations


def main():
    n, _ = read_numbers()
    binaries = read_binaries(n)
    print(*acm_team(binaries), sep='\n')


def read_numbers():
    return list(map(int, input().split()))


def read_binaries(n):
    return [int(input().strip(), 2) for _ in range(n)]


def acm_team(binaries):
    max_subjects = 0
    teams_with_max = 0

    for a, b in combinations(binaries, 2):
        known_subjects = (a | b).bit_count()

        if known_subjects > max_subjects:
            max_subjects = known_subjects
            teams_with_max = 1
        elif known_subjects == max_subjects:
            teams_with_max += 1

    return max_subjects, teams_with_max


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/40 - acm icpc team/acm_icpc_team.rb
# https://www.hackerrank.com/challenges/acm-icpc-team/problem?isFullScreen=true

def main
  n, _ = read_numbers
  binaries = read_binaries(n)
  puts acm_team(binaries)
end

def read_numbers
  gets.split.map!(&:to_i)
end

def read_binaries(n)
  n.times.map { gets.strip.to_i(2) }
end

def acm_team(binaries)
  max_subjects = 0
  teams_with_max = 0

  binaries.combination(2).each do |a, b|
    known_subjects = (a | b).to_s(2).count('1')

    if known_subjects > max_subjects
      max_subjects = known_subjects
      teams_with_max = 1
    elsif known_subjects == max_subjects
      teams_with_max += 1
    end
  end

  [max_subjects, teams_with_max]
end

main if __FILE__ == $0
